[7.x] Specify dynamic attribute casts at query-time

mpbarlow opened the pull request:
This pull request allows additional (or replacement) attribute casts to be specified on the Eloquent query builder, such that they apply only to the results of that query.

The primary use case for this functionality is to apply casts to columns resulting from subqueries or joins, that are not usually present on the model.

While adding the aliases for those columns to the `$casts` array on the model does work, I feel that polluting the model with columns that are simply not present most of the time is not ideal, especially when the actual definition of those column names are likely to be spread across many different files.

## Example Usage
Consider the following query selecting `User`s along with the date of their most recent `Post`:

```php
User
    ::select([
        'users.*',
        'last_posted_at' => Post::selectRaw('MAX(created_at)')
            ->whereColumn('user_id', 'users.id')
    ])
    ->get();
```

While `User`’s own timestamps will be cast to Carbon instances, `last_posted_at` will simply be a string. Currently, the only way to cast this would be to add `last_posted_at` to either `$casts` or `$dates` on `User`, despite the column only being present for this one specific query.

This PR allows the following:

```php
User
    ::select([
        'users.*',
        'last_posted_at' => Post::selectRaw('MAX(created_at)')
            ->whereColumn('user_id', 'users.id')
    ])
    ->withCasts(['last_posted_at' => 'date'])
    ->get();
```

where the array of casts will be merged into any existing casts specified on the model.

Naming-wise I’m not sure if “withCasts” is the best option; I’ve just gone along with the likes of “withCount”.

<!--
Please only send a pull request to branches which are currently supported: https://laravel.com/docs/releases#support-policy 

If you are unsure which branch your pull request should be sent to, please read: https://laravel.com/docs/contributions#which-branch

Pull requests without a descriptive title, thorough description, or tests will be closed.

In addition, please describe the benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.
-->

mpbarlow committed:
Specify dynamic attribute casts at query-time

This pull request allows additional (or replacement) attribute casts to be specified on the Eloquent query builder, such that they apply only to the results of that query.

The primary use case for this functionality is to apply casts to columns resulting from subqueries or joins, that are not usually present on the model.

While adding the aliases for those columns to the `$casts` array on the model does work, I feel that polluting the model with columns that are simply not present most of the time is not ideal, especially when the actual definition of those column names are likely to be spread across many different files.

Consider the following query selecting `User`s along with the date of their most recent `Post`:

```php
User
	::select([
		'users.*',
		'last_posted_at' => Post::selectRaw('MAX(created_at)')
			->whereColumn('user_id', 'users.id')
	])
	->get();
```

While `User`’s own timestamps will be cast to Carbon instances, `last_posted_at` will simply be a string. Currently, the only way to cast this would be to add `last_posted_at` to either `$casts` or `$dates` on `User`, despite the column only being present for this one specific query.

This PR allows the following:

```php
User
	::select([
		'users.*',
		'last_posted_at' => Post::selectRaw('MAX(created_at)')
			->whereColumn('user_id', 'users.id')
	])
	->withCasts(['last_posted_at' => 'date'])
	->get();
```

where the array of casts will be merged into any existing casts specified on the model.

Naming wise I’m not sure if “withCasts” is the best option; I’ve just gone along with the likes of “withCount”.

mfn commented:
If accepted, should already be forward-ported to support the Laravel 7 custom casts, see https://github.com/laravel/framework/pull/31035

GrahamCampbell commented:
Due to the changes in Laravel 7, I think this PR should only target Laravel 7.

mpbarlow commented:
> Due to the changes in Laravel 7, I think this PR should only target Laravel 7.

Makes sense to me. What’s the preferred way to do this, should I open a new PR for the 7.x branch and close this one?

GrahamCampbell commented:
It doesn't matter. You can rebase this against master, and update the target branch on GitHub, or just send a new PR.

mpbarlow commented:
I've rebased it onto master and updated the PR. 

It works nicely with the new custom cast changes. Everything happens before the models are hydrated, so I didn't end up needing to change anything.

taylorotwell commented:
Pretty nice, low-cost feature. Thanks! Actually could have used this myself not too long ago.
